File: examples/adaptive-slotbook/publication.py
"""Publication helpers for the finite binary qualification; no alternate execution path."""
import copy
import json
import secrets
import tomllib


def toml_document(document):
    """Write the concrete bootstrap configuration's strings, numbers, lists and tables."""
    lines = []

    def table(value, path=(), array=False):
        if path:
            name = '.'.join(json.dumps(part) for part in path)
            lines.append(('[' if not array else '[[') + name + (']' if not array else ']]'))
        for key, item in value.items():
            if not isinstance(item, dict) and not (isinstance(item, list) and item and isinstance(item[0], dict)):
                lines.append(json.dumps(key) + ' = ' + json.dumps(item))
        lines.append('')
        for key, item in value.items():
            if isinstance(item, dict):
                table(item, (*path, key))
            elif isinstance(item, list) and item and isinstance(item[0], dict):
                for entry in item:
                    table(entry, (*path, key), True)
    table(document)
    result = '\n'.join(lines)
    assert tomllib.loads(result) == document
    return result


def configure(text, root):
    config = tomllib.loads(text)
    config['runtime'].update(controller_activation='enabled', effect_threads=1, effect_queue=1,
                             global_concurrency=1, per_run_concurrency=1, per_capability_concurrency=1,
                             per_branch_concurrency=1, maximum_effect_claim=1,
                             publication_services={'method:slotbook': 'grant:operator'})
    config['serving'].update(worker_threads=1, observation_hot_retention_ms=100)
    config['serving']['clients']['execution_limits']['nested_invocations'] = {'process': 8, 'model': 0}
    operator = config['actors'][0]
    operator['authority']['resources']['capability']['identities']['values'].append('method:slotbook')
    operator['authority']['resources']['capability']['operations']['values'].extend(
        ['method.publish', 'method.inspect', 'method.retire', 'method.invoke'])
    consumer = copy.deepcopy(operator)
    consumer.update(actor='human:consumer', credential_ref='credential:consumer',
                    grant_id='grant:consumer', preset='invoker')
    resources = consumer['authority']['resources']
    resources['capability']['identities']['values'] = ['method:slotbook']
    resources['capability']['operations']['values'] = ['method.invoke']
    resources['artifacts'] = {'type': 'deny_all'}
    resources['filesystem'] = []
    resources['workspace'] = {'allow_any_in_run': False, 'scopes': []}
    resources['secrets'] = []
    config['actors'].append(consumer)
    credential = root / 'consumer.token'
    credential.write_text(secrets.token_hex(32))
    credential.chmod(0o600)
    config['secret_sources']['credential:consumer'] = {'type': 'file', 'path': str(credential)}
    return toml_document(config)


def method(descriptor, revision, authority):
    descriptor = copy.deepcopy(descriptor)
    descriptor.update(identity='method:slotbook', descriptor_revision=1, category={'type': 'tool'})
    descriptor['operations'] = {'method.invoke': descriptor['operations']['workspace.execute']}
    return {
        'schema_version': 1, 'descriptor': descriptor,
        'documentation': 'Verify and deploy the governed Slotbook candidate to the fixed test installation. '
                         'A successful invocation confirms the internal agreement; application service state remains separately inspectable.',
        'revision': revision['id'], 'agreement': revision['semantic']['agreement']['digest'],
        'service': {'actor': authority['actor'], 'grant': authority['grant_id'],
                    'grant_revision': authority['grant_revision'], 'grant_digest': authority['grant_digest'],
                    'revocation_generation': 0},
        'inputs': {}, 'outputs': {},
        'workspace_budget': {'max_value_versions': 1024, 'max_inline_bytes_per_value': 65536,
                             'max_total_inline_bytes': 1048576, 'max_artifacts': 256,
                             'max_bytes_per_artifact': 16777216, 'max_total_artifact_bytes': 67108864},
        'allowance': {'cost_micros': 0, 'currency': None, 'input_units': 0, 'output_units': 0,
                      'artifact_bytes': 33554432, 'process_admissions': 8, 'model_admissions': 0},
        'maximum_outstanding': 1, 'maximum_depth': 4, 'maximum_duration_ms': 240000,
    }


def outer_method(root, cli, capability, peer=None):
    """Author a caller workflow through the supported CLI with an explicit provider selection."""
    import subprocess
    internal = json.loads((root / 'governed.json').read_text())['revision']
    task = copy.deepcopy(internal['semantic']['nodes']['repair.begin'])
    task.update(id='invoke-method', control_inputs=[], data_inputs={}, data_outputs={})
    requirement = task['kind']['config']['requirement']
    requirement.update(categories=[{'type': 'tool'}], exact_capability=capability, operation='method.invoke')
    if peer:
        requirement['placement'] = {'localities': ['peer'], 'peers': [peer]}
    done = copy.deepcopy(internal['semantic']['nodes']['done'])
    mutations = [{'type': 'add_node', 'node': task}, {'type': 'add_node', 'node': done},
                 {'type': 'add_edge', 'edge': {'id': 'finish', 'kind': 'control',
                  'source_node': 'invoke-method', 'source_port': 'out', 'target_node': 'done', 'target_port': 'in'}}]
    path = root / 'outer-mutations.json'
    path.write_text(json.dumps(mutations))
    output = root / 'outer.json'
    subprocess.run([cli, '--json', 'blueprint', 'create', str(path), '--workflow', 'slotbook-caller',
                    '--author', 'agent:repair', '--output', str(output)], capture_output=True, text=True, check=True)
    return json.loads(output.read_text())['revision']


def peer_configuration(text, root, provider_port):
    """Two explicit loopback hosts, separate stores, fixed peer credentials and one worker each."""
    config = tomllib.loads(text)
    origin_port = provider_port + 1
    credential = root / 'peer.token'
    credential.write_text(secrets.token_hex(32))
    credential.chmod(0o600)
    config['secret_sources']['credential:peer'] = {'type': 'file', 'path': str(credential)}

    def relationship(peer, port):
        return {'peer_id': peer, 'endpoint': f'http://127.0.0.1:{port}/',
                'credential_ref': 'credential:peer', 'insecure_loopback_development': True,
                'minimum_minor': 5, 'maximum_minor': 5,
                'actions': ['read_catalog', 'invoke', 'cancel', 'artifact_upload', 'artifact_download'],
                'capability_allow': ['method:slotbook'], 'capability_deny': [],
                'operation_allow': ['method.invoke'], 'maximum_side_effect': 'non_idempotent_write',
                'maximum_concurrent': 1, 'maximum_requests_per_minute': 6000,
                'maximum_artifact_bytes': 67108864, 'artifact_sensitivities': ['internal', 'restricted'],
                'maximum_duration_ms': 240000, 'maximum_cost_micros': 0,
                'maximum_input_units': 0, 'maximum_output_units': 0,
                'nested_invocations': {'process': 8, 'model': 0},
                'maximum_observations': 256, 'catalog_ttl_ms': 300000,
                'trust_zone': 'slotbook-qualification', 'delegation_ref': 'delegation:slotbook',
                'expires_at_unix_ms': 4102444800000, 'enabled': True}

    config['peers'] = {'mode': 'enabled', 'relationships': [relationship('host:slotbook-caller', origin_port)]}
    origin = copy.deepcopy(config)
    origin.update(host_id='host:slotbook-caller', data_root=str(root / 'origin-data'),
                  bind=f'127.0.0.1:{origin_port}')
    origin['runtime']['publication_services'] = {}
    origin['adapters'] = {'process_profiles': [], 'model_profiles': []}
    origin['peers']['relationships'] = [relationship('host:slotbook-test', provider_port)]
    origin['actors'] = [origin['actors'][0]]
    resources = origin['actors'][0]['authority']['resources']
    resources['capability']['identities'] = {'type': 'any'}
    resources['capability']['operations'] = {'type': 'only', 'values': ['method.invoke']}
    resources['filesystem'] = []
    resources['network'] = {'profiles': ['peer:host:slotbook-test'], 'destinations': [f'127.0.0.1:{provider_port}']}
    resources['peers'] = {'allow_any': False, 'identities': ['host:slotbook-test']}
    origin_path = root / 'origin.toml'
    origin_path.write_text(toml_document(origin))
    return toml_document(config), origin_path, origin_port

File: tools/evidence/src/bin/slotbook-evidence/publication.rs
use super::prepare::{self, INTERNAL_ARTIFACT_BYTES};
use milkdrift_evidence::{EvidenceResult, application::write_private};
use serde_json::{Value, json};
use std::path::{Path, PathBuf};

pub(super) fn configure(config: &mut Value, root: &Path) -> EvidenceResult {
    let runtime = config["runtime"]
        .as_object_mut()
        .ok_or("runtime config absent")?;
    for key in [
        "effect_threads",
        "effect_queue",
        "global_concurrency",
        "per_run_concurrency",
        "per_capability_concurrency",
        "per_branch_concurrency",
        "maximum_effect_claim",
    ] {
        runtime.insert(key.into(), json!(1));
    }
    runtime.insert("controller_activation".into(), json!("enabled"));
    runtime.insert(
        "publication_services".into(),
        json!({"method:slotbook":"grant:operator"}),
    );
    config["serving"]["worker_threads"] = json!(1);
    config["serving"]["observation_hot_retention_ms"] = json!(100);
    config["serving"]["clients"]["execution_limits"]["nested_invocations"] =
        json!({"process":8,"model":0});
    config["serving"]["clients"]["execution_limits"]["artifact_bytes"] =
        json!(INTERNAL_ARTIFACT_BYTES);
    let operator = &mut config["actors"][0];
    operator["authority"]["resources"]["capability"]["identities"]["values"]
        .as_array_mut()
        .ok_or("capability identities absent")?
        .push(json!("method:slotbook"));
    operator["authority"]["resources"]["capability"]["operations"]["values"]
        .as_array_mut()
        .ok_or("capability operations absent")?
        .extend(
            [
                "method.publish",
                "method.inspect",
                "method.retire",
                "method.invoke",
            ]
            .map(|v| json!(v)),
        );
    let mut consumer = operator.clone();
    consumer["actor"] = json!("human:consumer");
    consumer["credential_ref"] = json!("credential:consumer");
    consumer["grant_id"] = json!("grant:consumer");
    consumer["preset"] = json!("invoker");
    let resources = &mut consumer["authority"]["resources"];
    resources["capability"]["identities"]["values"] = json!(["method:slotbook"]);
    resources["capability"]["operations"]["values"] = json!(["method.invoke"]);
    resources["artifacts"] = json!({"type":"deny_all"});
    resources["filesystem"] = json!([]);
    resources["workspace"] = json!({"allow_any_in_run":false,"scopes":[]});
    resources["secrets"] = json!([]);
    config["actors"]
        .as_array_mut()
        .ok_or("actors absent")?
        .push(consumer);
    let credential = root.join("consumer.token");
    prepare::credential(&credential)?;
    config["secret_sources"]["credential:consumer"] = json!({"type":"file","path":credential});
    Ok(())
}
pub(super) fn method(mut descriptor: Value, revision: &Value, authority: &Value) -> Value {
    descriptor["identity"] = json!("method:slotbook");
    descriptor["descriptor_revision"] = json!(1);
    descriptor["category"] = json!({"type":"tool"});
    descriptor["operations"] =
        json!({"method.invoke":descriptor["operations"]["workspace.execute"]});
    json!({"schema_version":1,"descriptor":descriptor,"documentation":"Verify and deploy the governed Slotbook candidate to the fixed test installation. A successful invocation confirms the internal agreement; application service state remains separately inspectable.","revision":revision["id"],"agreement":revision["semantic"]["agreement"]["digest"],"service":{"actor":authority["actor"],"grant":authority["grant_id"],"grant_revision":authority["grant_revision"],"grant_digest":authority["grant_digest"],"revocation_generation":0},"inputs":{},"outputs":{},"workspace_budget":{"max_value_versions":1024,"max_inline_bytes_per_value":65536,"max_total_inline_bytes":1048576,"max_artifacts":256,"max_bytes_per_artifact":33554432,"max_total_artifact_bytes":INTERNAL_ARTIFACT_BYTES},"allowance":{"cost_micros":0,"currency":null,"input_units":0,"output_units":0,"artifact_bytes":INTERNAL_ARTIFACT_BYTES,"process_admissions":8,"model_admissions":0},"maximum_outstanding":1,"maximum_depth":4,"maximum_duration_ms":240000})
}
pub(super) fn outer(
    root: &Path,
    cli: &Path,
    capability: &str,
    peer: bool,
) -> EvidenceResult<Value> {
    let document: Value = serde_json::from_slice(&std::fs::read(root.join("governed.json"))?)?;
    let internal = &document["revision"];
    let mut task = internal["semantic"]["nodes"]["repair.begin"].clone();
    task["id"] = json!("invoke-method");
    task["control_inputs"] = json!([]);
    task["data_inputs"] = json!({});
    task["data_outputs"] = json!({});
    let req = &mut task["kind"]["config"]["requirement"];
    req["categories"] = json!([{"type":"tool"}]);
    req["exact_capability"] = json!(capability);
    req["operation"] = json!("method.invoke");
    if peer {
        req["placement"] = json!({"localities":["peer"],"peers":["host:slotbook-test"]});
    }
    let mutations = json!([{"type":"add_node","node":task},{"type":"add_node","node":internal["semantic"]["nodes"]["done"]},{"type":"add_edge","edge":prepare::edge("finish","invoke-method","done","control","out","in")}]);
    let input = prepare::write(root, "outer-mutations.json", &mutations)?;
    prepare::local(
        cli,
        &[
            "blueprint".into(),
            "create".into(),
            prepare::argument(input),
            "--workflow".into(),
            "slotbook-caller".into(),
            "--author".into(),
            "agent:repair".into(),
            "--output".into(),
            prepare::argument(root.join("outer.json")),
        ],
    )?;
    Ok(
        serde_json::from_slice::<Value>(&std::fs::read(root.join("outer.json"))?)?["revision"]
            .clone(),
    )
}
pub(super) fn peer_configuration(
    config: &mut Value,
    root: &Path,
    provider_port: u16,
) -> EvidenceResult<(PathBuf, u16)> {
    let origin_port = provider_port.checked_add(1).ok_or("peer port overflows")?;
    let credential = root.join("peer.token");
    prepare::credential(&credential)?;
    config["secret_sources"]["credential:peer"] = json!({"type":"file","path":credential});
    let relationship = |peer: &str, port: u16| json!({"peer_id":peer,"endpoint":format!("http://127.0.0.1:{port}/"),"credential_ref":"credential:peer","insecure_loopback_development":true,"minimum_minor":5,"maximum_minor":5,"actions":["read_catalog","invoke","cancel","artifact_upload","artifact_download"],"capability_allow":["method:slotbook"],"capability_deny":[],"operation_allow":["method.invoke"],"maximum_side_effect":"non_idempotent_write","maximum_concurrent":1,"maximum_requests_per_minute":6000,"maximum_artifact_bytes":INTERNAL_ARTIFACT_BYTES,"artifact_sensitivities":["internal","restricted"],"maximum_duration_ms":240000,"maximum_cost_micros":0,"maximum_input_units":0,"maximum_output_units":0,"nested_invocations":{"process":8,"model":0},"maximum_observations":256,"catalog_ttl_ms":300000,"trust_zone":"slotbook-qualification","delegation_ref":"delegation:slotbook","expires_at_unix_ms":4102444800000_u64,"enabled":true});
    config["peers"] = json!({"mode":"enabled","relationships":[relationship("host:slotbook-caller",origin_port)]});
    let mut origin = config.clone();
    origin["host_id"] = json!("host:slotbook-caller");
    origin["data_root"] = json!(root.join("origin-data"));
    origin["bind"] = json!(format!("127.0.0.1:{origin_port}"));
    origin["runtime"]["publication_services"] = json!({});
    origin["adapters"] = json!({"process_profiles":[],"model_profiles":[]});
    origin["peers"]["relationships"] = json!([relationship("host:slotbook-test", provider_port)]);
    origin["actors"] = json!([origin["actors"][0]]);
    let resources = &mut origin["actors"][0]["authority"]["resources"];
    resources["capability"]["identities"] = json!({"type":"any"});
    resources["capability"]["operations"] = json!({"type":"only","values":["method.invoke"]});
    resources["filesystem"] = json!([]);
    resources["network"] = json!({"profiles":["peer:host:slotbook-test"],"destinations":[format!("127.0.0.1:{provider_port}")]});
    resources["peers"] = json!({"allow_any":false,"identities":["host:slotbook-test"]});
    let path = root.join("origin.toml");
    write_private(&path, toml::to_string_pretty(&origin)?.as_bytes())?;
    Ok((path, origin_port))
}
